proc/groupby.go

```diff
@@ -353,56 +502,151 @@ func (g *GroupByAggregator) Consume(r *zng.Record) error {
 	return nil
 }
 
-func (g *GroupByAggregator) updateMaxKey(v zng.Value) {
-	if g.maxKey == nil {
-		g.maxKey = &v
+func (g *GroupByAggregator) spillTable(eof bool) error {
+	parts, err := g.readTable(true, true)
+	if err != nil {
+		return err
+	}
+	if parts == nil {
+		return nil
+	}
+	recs := parts.Records()
+	expr.SortStable(recs, g.keysCompare)
+
+	if !eof && g.inputSortDir != 0 {
+		keyVal, err := g.keys[0].expr(recs[len(recs)-1])
+		if err != nil && !errors.Is(err, zng.ErrUnset) {
+			return err
+		}
+		g.updateMaxSpillKey(keyVal)
+	}
+
+	spill, err := g.spillManager.writeSpill(g.zctx, parts)
+	if err != nil {
+		return err
+	}
+	g.combiner.AddReader(spill)
+	return nil
+}
+
+func (g *GroupByAggregator) updateMaxTableKey(v zng.Value) {
+	if g.maxTableKey == nil {
+		g.maxTableKey = &v
 		return
 	}
-	if g.valueCompare(v, *g.maxKey) > 0 {
-		g.maxKey = &v
+	if g.valueCompare(v, *g.maxTableKey) > 0 {
+		g.maxTableKey = &v
 	}
 }
 
+func (g *GroupByAggregator) updateMaxSpillKey(v zng.Value) {
+	if g.maxSpillKey == nil {
+		g.maxSpillKey = &v
+		return
+	}
+	if g.valueCompare(v, *g.maxSpillKey) > 0 {
+		g.maxSpillKey = &v
+	}
+}
+
+func (g *GroupByAggregator) haveSpills() bool {
+	return g.spillManager.n > 0
+}
+
 // Results returns a batch of aggregation result records. If the input
-// is sorted in the primary grouping key, this can be called multiple times;
-// all completed keys at the time of the invocation are returned (but
-// not necessarily in their input sort order). A final call with
-// eof=true should be made to get the final keys.
-//
-// If the input is not sorted, a single call (with eof=true) should be
-// made after all records have been Consumed()'d.
+// is sorted in the primary key, only keys that are completed are
+// returned (but not necessarily in their input sort order). A final
```

Comment:
`... not necessarily in their input sort order`: I see that g.keycompare is created based on the input sort direction, and it's used to sort the output per-batch records, so it seems like this comment isn't accurate? Or am I misunderstanding?

Reply:
say you're doing `every 1s count() by ip`, and the input is sorted in `ts`, then what i was trying to express here is that the output will also be sorted in `ts`, but the output order of the ips may not match the input order. For example, with this input:

```
ts:1, ip:1.1.1.1
ts:1, ip:1.1.1.2
ts:2, ip:1.1.1.1
```

you might get this output where 1.1.1.2 is ahead of 1.1.1.1:

```
ts:1, count:1, ip:1.1.1.2
ts:1, count:1, ip:1.1.1.1
ts:2, count:1, ip:1.1.1.1
```

Maybe I should just drop the `(but not necessarily in their input sort order)` from the comment as I now I think of it I don't expect one would have any sortedness expectations here?

Reply:
ah, thanks for the clarification; yes, I think you can just drop that portion.